### PR TITLE
Support .NET 7.0

### PR DESCRIPTION
**Pull request by zcsizmadia**
.NET 7.0 just went RC (rc.1) and expected to be officially released in mid November 2022. This PR should be merged once .NET 7.0 is released on November, however in the meantime the build process can be validated.

Note:
Before merging `include-prerelease: true` lines should be removed from the github workflow `actions/setup-dotnet@v2` actions.

### CODE REVIEW

File: src/GraphQL.ApiTests/net6+net7/GraphQL.approved.txt
Comment (zcsizmadia) [inline src/GraphQL.ApiTests/net6+net7/GraphQL.approved.txt:1180]: Need advice with this change

File: src/GraphQL.ApiTests/netstandard20+netstandard21/GraphQL.approved.txt
Comment (zcsizmadia) [inline src/GraphQL.ApiTests/netstandard20+netstandard21/GraphQL.approved.txt:1180]: Need advice with this change

**Comment by zcsizmadia** [comment]
.NET Core 3.1 will go EOL mid December 2022 and .NET 5.0 is already EOL (https://dotnet.microsoft.com/en-us/download/dotnet)

Those targets could be removed as well when NET 7 is released. Supporting those EOL frameworks will be ok using the netstandard2.x targets.

File: src/GraphQL.Tests/GraphQLValuesCacheTests.cs
Comment (zcsizmadia) [inline src/GraphQL.Tests/GraphQLValuesCacheTests.cs:16]: It is unfortunate that MS has changed the FormatException message from .NET 6 -> .NET 7. Maybe the strict exception message check could be removed and just check for throwing `FormatException`. In that case the nasty ifdef could be eliminated.

File: src/GraphQL.Tests/GraphQLValuesCacheTests.cs
Comment (zcsizmadia) [inline src/GraphQL.Tests/GraphQLValuesCacheTests.cs:34]: It is unfortunate that MS has changed the FormatException message from .NET 6 -> .NET 7. Maybe the strict exception message check could be removed and just check for throwing `FormatException`. In that case the nasty `ifdef` could be eliminated.

File: src/GraphQL.Tests/Instrumentation/ApolloTracingTests.cs
Comment (zcsizmadia) [inline src/GraphQL.Tests/Instrumentation/ApolloTracingTests.cs:71]: To eliminate the framework specific expected string, can the `ms` fraction be changed to `102`?  In that case the formatting would match all .NET frameworks.

Comment (Shane32) [reply src/GraphQL.Tests/Instrumentation/ApolloTracingTests.cs:71]: Let's leave it as you have it now.  I would prefer to investigate further and determine why the change occurred, and then determine if we should change the serializer implementation to format the value in a defined manner regardless of the target framework.

File: src/GraphQL/GraphQL.csproj
Comment (Shane32) [inline src/GraphQL/GraphQL.csproj:4]: Please remove .NET 7 TFM until we have a ifdef that requires it.

File: src/GraphQL.SystemTextJson/GraphQL.SystemTextJson.csproj
Comment (zcsizmadia) [reply src/GraphQL.SystemTextJson/GraphQL.SystemTextJson.csproj:4]: Adding net7 there will use the STJ 7.0+, instead of (4.7,2) which is used in case of netstandard2.x target. MS claims performance gain in STJ 7 compared to earlier version.

File: src/GraphQL/GraphQL.csproj
Comment (zcsizmadia) [reply src/GraphQL/GraphQL.csproj]: Removed.

**Comment by Shane32** [comment]
> .NET Core 3.1 will go EOL mid December 2022 and .NET 5.0 is already EOL (https://dotnet.microsoft.com/en-us/download/dotnet)
> 
> Those targets could be removed as well when NET 7 is released. Supporting those EOL frameworks will be ok using the netstandard2.x targets.

GraphQL.NET only targets .NET Standard 2.0, .NET Standard 2.1 and .NET 6, as there are no ifdefs for other target frameworks.  System.Text.Json includes a target for .NET Core 3.1 because System.Text.Json is included with .NET Core 3.0+.  As such, the dependency can be eliminated.

At present I see no reason to change the targeted frameworks.  I feel that .NET Core 3.1 is especially important because not only it is a LTS release, a number of the packages developed at that time are cross-platform compatible with .NET Framework -- for example, Entity Framework Core.  While perhaps these applications can all run on .NET 6, when using the 3.1.x components of that time, it may be more likely that applications written for .NET Core 3.1 designed to be cross compatible with .NET Framework may not yet be upgraded to .NET 6.  There is a similar situation that exists with .NET Core 2.1 for web applications -- hence why GraphQL.NET Server 7.x now supports .NET Framework and .NET Core 2.1 web applications.  (Not that in that situation, there is official support from Microsoft for ASP.NET Core 2.1.x when run on supported editions of .NET Framework.)

I would have little argument against removing .NET 5 code -- it was never a LTS release and is more of a preview of .NET 6, just like .NET Core 3.0 was compared to .NET Core 3.1.  But yet, there is no reason to at this time.

If and when it becomes troublesome to perform development/testing for older frameworks (e.g. trying to develop/test .NET Framework 4.5 applications today), we will likely drop those frameworks from the test suite and/or libraries at that time.

Thanks for the PR!  As you suggest, we likely will not merge the PR until .NET 7.0 is released.

File: src/GraphQL.SystemTextJson/GraphQL.SystemTextJson.csproj
Comment (Shane32) [reply src/GraphQL.SystemTextJson/GraphQL.SystemTextJson.csproj:4]: System.Text.Json is not a dependency for .NET Core 3.1+ as it is released as part of the framework.  .NET 7 will use the .NET 6 TFM as it is the closest match.  As such, it should have all of the performance benefits available with .NET 7.

File: src/GraphQL.Tests/Instrumentation/ApolloTracingTests.cs
Comment (zcsizmadia) [reply src/GraphQL.Tests/Instrumentation/ApolloTracingTests.cs:71]: It seems to be the issue is in the DateTime formatting library:

```
DateTime dt = new DateTime(2019, 12, 05, 15, 38, 00, DateTimeKind.Utc);
dt = dt.AddMilliseconds(102.5);
Console.WriteLine($"{dt.ToString("O")}");
```

net 7.0 -> `2019-12-05T15:38:00.1025000Z`
net5.0, net6.0 -> `2019-12-05T15:38:00.1030000Z`

Comment (Shane32) [reply src/GraphQL.Tests/Instrumentation/ApolloTracingTests.cs:71]: Or perhaps `AddMilliseconds` does not retain the fractional portion of the milliseconds specified... ?

- https://github.com/dotnet/runtime/pull/73198

Looks like this line here:

```cs
return AddTicks((long)millis_double * TicksPerMillisecond);
```

should have been this:

```cs
return AddTicks((long)(millis_double * TicksPerMillisecond));
```

but they cleaned it all up for .NET 7.

Comment (Shane32) [reply src/GraphQL.Tests/Instrumentation/ApolloTracingTests.cs:71]: So we could ignore it, or we could change the call to `AddMilliseconds` to use `AddTicks` instead, which would fix the bug in .NET 6 and prior versions.

File: src/GraphQL.SystemTextJson/GraphQL.SystemTextJson.csproj
Comment (zcsizmadia) [reply src/GraphQL.SystemTextJson/GraphQL.SystemTextJson.csproj:4]: Roger. I always assumed. If I build a net 6.0 app  always the STJ latest 6.0.x will be package referenced and not the absolute latest (in this case 7.0.x). However I never tested that, so  I will remove net7 as suggested

Comment (zcsizmadia) [reply src/GraphQL.SystemTextJson/GraphQL.SystemTextJson.csproj]: Removed net7

File: src/GraphQL.Tests/Instrumentation/ApolloTracingTests.cs
Comment (Shane32) [reply src/GraphQL.Tests/Instrumentation/ApolloTracingTests.cs]: :+1:

File: src/GraphQL.ApiTests/net6+net7/GraphQL.approved.txt
Comment (sungam3r) [reply src/GraphQL.ApiTests/net6+net7/GraphQL.approved.txt]: @Shane32  I think there are no problems if we all move to > 6.0.402 SDK . @zcsizmadia Please rebase on `master`. I removed these lines in #3380 .

**Comment by sungam3r** [comment]
> Those targets could be removed as well when NET 7 is released. Supporting those EOL frameworks will be ok using the netstandard2.x targets.

Well, I'm not so positive to quickly remove support for "old" TFMs especially if this support is given at low price.

File: src/GraphQL.Tests/GraphQLValuesCacheTests.cs
Comment (sungam3r) [reply src/GraphQL.Tests/GraphQLValuesCacheTests.cs:16]: > It is unfortunate that MS has changed the FormatException message from .NET 6 -> .NET 7. 

Maybe yes, maybe no :) . Yes, I prefer to test exception messages, not only exception types. When I read the test code, I better understand the nature of the code tested. I would not call it a tough requirement, just a recommendation that I would line to adhere to.

> So we just update the tests - as you have already done - when necessary.

Yes. Update if necessary. Sometimes I use `StartsWith`/`Contains` instead of exact match.

File: src/GraphQL/Instrumentation/ApolloTrace.cs
Comment (sungam3r) [inline src/GraphQL/Instrumentation/ApolloTrace.cs:18]: 👍

**Comment by sungam3r** [comment]
> I would have little argument against removing .NET 5 code -- it was never a LTS release and is more of a preview of .NET 6, just like .NET Core 3.0 was compared to .NET Core 3.1. But yet, there is no reason to at this time.

> If and when it becomes troublesome to perform development/testing for older frameworks (e.g. trying to develop/test .NET Framework 4.5 applications today), we will likely drop those frameworks from the test suite and/or libraries at that time.

> Thanks for the PR! As you suggest, we likely will not merge the PR until .NET 7.0 is released.

Agree. Let's merge ASAP after NET7 release.

File: .github/workflows/build-artifacts-code.yml
Comment (sungam3r) [inline .github/workflows/build-artifacts-code.yml:42]:
```suggestion
      - name: Setup .NET SDK
```

File: .github/workflows/codeql-analysis.yml
Comment (sungam3r) [inline .github/workflows/codeql-analysis.yml:19]:
```suggestion
    - name: Setup .NET SDK
```

File: .github/workflows/publish-code.yml
Comment (sungam3r) [inline .github/workflows/publish-code.yml:49]:
```suggestion
      - name: Setup .NET SDK
```

**Review by sungam3r — CHANGES_REQUESTED** [review]
I'm fine to merge this one after resolving conflicts and .NET7 releaase.

**Comment by zcsizmadia** [comment]
@sungam3r Merged master into the PR:

1. Using EOL SDKs the build will generat several EOL warnings. As of now it is only 5.0, however 3.1 goes EOL soon, so that will do th same. Added `<CheckEolTargetFramework>false</CheckEolTargetFramework>` to Directory.build.props to disable disable that warning. (EOL SDKs are used only for tests and benchmarks)
2. `dotnet-quality: preview` should be removed from the github workflow yml files before merging this PR

**Comment by sungam3r** [comment]
`CheckEolTargetFramework` is fine, we use it in the server project as well.

**Review by sungam3r — APPROVED** [review]

File: src/Directory.Build.props
Comment (Shane32) [inline src/Directory.Build.props:32]:
```suggestion
```

I would remove this comment; it is false.  For example, GraphQL.SystemTextJson is also compiled for .NET 5, and GraphQL.Server is also compiled for .NET Core 2.1.

Besides, the comment should be unnecessary -- the property name is self descriptive and can be easily looked up if not.

File: src/GraphQL/Instrumentation/ApolloTrace.cs
Comment (Shane32) [reply src/GraphQL/Instrumentation/ApolloTrace.cs:18]: This should be patched in a separate PR.  While perhaps not likely to matter, it can/will cause different behavior in the GraphQL runtime.

Comment (Shane32) [reply src/GraphQL/Instrumentation/ApolloTrace.cs:18]: - #3384

**Comment by zcsizmadia** [comment]
@Shane32 
1. Removed the comments for CheckEolTargetFramework
2. Removed the ApolloTrace AddTicks fix. The tests in this PR will fail until #3384 is not merged.

File: src/GraphQL/Instrumentation/ApolloTrace.cs
Comment (zcsizmadia) [reply src/GraphQL/Instrumentation/ApolloTrace.cs]: Done

File: src/Directory.Build.props
Comment (zcsizmadia) [reply src/Directory.Build.props]: Done

**Review by Shane32 — APPROVED** [review]

**Comment by zcsizmadia** [comment]
NEt 7 was released today. Merged master into the PR branch and removed the dotnet-quality=preview lines from the github workflows.

**Comment by Shane32** [comment]
Visual Studio 17.4 hasn't been released yet.  Probably it will be within a day or so and then we can merge this.

**Comment by sungam3r** [comment]
I have run VS installer just now - 2.4 Gb update for 17.4 :)

**Comment by sungam3r** [comment]
Successful build on my machine except broken localization in console
изображение

**Pull request merged**
